File: C++/331.cpp
__________________________________________________________________________________________________
sample 4 ms submission
class Solution {
public:
    bool isValidSerialization(string preorder) {
        int len = preorder.size();
        if(len == 0)
            return true;
        int n = 0;
        for(int i = len-1; i >= 0; i--){
            if(preorder[i] == ',')
                continue;
            else if(preorder[i] == '#')
                n++;
            else{
                if(n >= 2){
                    n--;
                    while(i >= 0 && preorder[i] != ',')
                        i--;
                }
                else
                    return false;
            }
        }
        if(n == 1)
            return true;
        else
            return false;
    }
};
__________________________________________________________________________________________________
sample 8508 kb submission
class Solution {
public:
    bool isValidSerialization(string preorder) {
        if(preorder.size() == 0)
            return false;
        
        int spaces = 1;
        
        for(int i = 0; i < preorder.size(); ++i) {
            spaces--;

            if(spaces < 0)
                return false;
            
            if(preorder[i] != '#')
                spaces += 2;
            
            while(i < preorder.size() && preorder[i] != ',')
                i++;
        }
        
        if(spaces == 0)
            return true;
        
        return false;
    }
};
__________________________________________________________________________________________________

File: Python3/331.py
__________________________________________________________________________________________________
sample 20 ms submission
class Solution:
    def isValidSerialization(self, preorder: str) -> bool:
        d=1;
        preorder=preorder.split(',')
        for i in range(len(preorder)):
            if preorder[i]=='#':
                d-=1
            else:
                d+=1
                
            if d<0 or (d==0 and i!=len(preorder)-1):
                return False
        if d==0:
            return True
        return False
__________________________________________________________________________________________________
sample 13040 kb submission
class Solution:
    def isValidSerialization(self, preorder: str) -> bool:
        
        need = 1
        for val in preorder.split(','):
            if not need:
                return False
            if val == '#':
                need -= 1
            else: 
                need +=1
        return not need
__________________________________________________________________________________________________
